### PR TITLE
pkg/operator: lay down ClusterOperator for MCO

### PR DESCRIPTION
**Pull request by runcom**
Closes https://github.com/openshift/machine-config-operator/issues/383

Signed-off-by: Antonio Murdaca <runcom@linux.com>

### CODE REVIEW

**Comment by runcom** [comment]
Also, I'm still learning on how to deploy this kind of changes from the installer to test all of this out (I'd love any guidance if you have time :angel: )

**Comment by abhinavdahiya** [comment]
The operator needs to create and mange the clusteroperator object.

https://github.com/openshift/cluster-version-operator/blob/master/docs/dev/clusteroperator.md

**Comment by bparees** [comment]
this repo needs an e2e-aws job so you don't break e2e-aws w/ a merge here.

**Comment by runcom** [comment]
> this repo needs an e2e-aws job so you don't break e2e-aws w/ a merge here.

probably a Prow issue since other PRs are running it

**Comment by bparees** [comment]
> The operator needs to create and mange the clusteroperator object.

the operator doesn't need to create it because the CVO is going to create it from the manifest, no?

as for managing it, i assumed this operator logic already was updating a clusteroperator object, but sure if it's not that obviously needs to happen too.. without that, the PR will never pass e2e-aws.

**Comment by cgwalters** [comment]
@abhinavdahiya this came out of https://url.corp.redhat.com/dfe5ca9

**Comment by abhinavdahiya** [comment]
If you read https://github.com/openshift/cluster-version-operator/blob/master/docs/dev/clusteroperator.md#how-should-i-include-clusteroperator-custom-resource-in-manifests

It clearly states in the NOTE  `Also note that the ClusterOperator resource in /manifests is only a communication mechanism, to tell the ClusterVersionOperator, which ClusterOperator resource to wait for. The ClusterVersionOperator does not create the ClusterOperator resource, this and updating it is the responsibility of the respective operator.`

**Comment by bparees** [comment]
> It clearly states in the NOTE `Also note that the ClusterOperator resource in /manifests is only a communication mechanism, to tell the ClusterVersionOperator, which ClusterOperator resource to wait for.`

that strikes me as a strange decision.  Everything else in manifests is explicitly created/applied by the CVO, correct?  Why not also create the clusteroperator resource?  (obviously it would be empty, but so what?).  Why the special case?

(since our operator started out creating the clusteroperator programmatically, and i'm guessing this operator does the same, it's probably moot, but it's likely to catch someone out at some point).

**Comment by runcom** [comment]
> OK so if we're in agreement that installs should gate on our operator, why is that not happening today and how do we fix it?

my understanding is that we still need to ship that CO yaml so the CVO can monitor it and gate on us

**Comment by runcom** [comment]
we may be hitting some aws limit on e2e I guess

/test e2e-aws

**Comment by runcom** [comment]
pushed a dummy commit to check tests are failing if we fail to report Available status

**Comment by runcom** [comment]
Alright removing WIP, seems like now we're gating at installer:

```
time="2019-02-07T12:06:51Z" level=debug msg="Still waiting for the cluster to initialize: Cluster operator machine-config has not yet reported success"
time="2019-02-07T12:10:36Z" level=debug msg="Still waiting for the cluster to initialize: Cluster operator machine-config is reporting a failure: Failed when progressing towards 3.11.0-584-gee2f741f-dirty because: error syncing: [dummy, timed out waiting for the condition during syncRequiredMachineConfigPools: pool master has not progressed to latest configuration: configuration for pool master is empty]"
```

**Comment by runcom** [comment]
It's also catching my dummy error, even this is going to fail in master currently cause we need #385

@cgwalters ptal

**Comment by runcom** [comment]
still on hold cause of #385 since masters are degraded anyway in master right now

**Comment by runcom** [comment]
Picked Colin's #385 here to check  if everything is green

**Comment by runcom** [comment]
Failing on:
```
time="2019-02-07T13:27:45Z" level=debug msg="Still waiting for the cluster to initialize: Cluster operator machine-config has not yet reported success"
time="2019-02-07T13:31:30Z" level=debug msg="Still waiting for the cluster to initialize: Cluster operator machine-config is reporting a failure: Failed when progressing towards 3.11.0-584-g5cef4210-dirty because: error syncing: [timed out waiting for the condition during waitForControllerConfigToBeCompleted: controllerconfig is not completed: status for ControllerConfig machine-config-controller is being reported for 0, expecting it for 1, timed out waiting for the condition during syncRequiredMachineConfigPools: pool master has not progressed to latest configuration: configuration for pool master is empty]"
time="2019-02-07T13:55:02Z" level=fatal msg="failed to initialize the cluster: timed out waiting for the condition"
```
Checking

**Comment by runcom** [comment]
actually, from logs, we're still getting #367 

```
 36 I0207 13:30:46.538517   12575 daemon.go:894] While getting MachineConfig master-cceb2fa93fab8800cb9e83839a1d0142, got: machineconfigs.machineconfiguration.openshift.io "master-cceb2fa93fab8800cb9e83839a1d0142" not found. Retrying...
 37 I0207 13:30:56.533566   12575 daemon.go:894] While getting MachineConfig master-cceb2fa93fab8800cb9e83839a1d0142, got: machineconfigs.machineconfiguration.openshift.io "master-cceb2fa93fab8800cb9e83839a1d0142" not found. Retrying...
 38 I0207 13:31:06.541155   12575 daemon.go:894] While getting MachineConfig master-cceb2fa93fab8800cb9e83839a1d0142, got: machineconfigs.machineconfiguration.openshift.io "master-cceb2fa93fab8800cb9e83839a1d0142" not found. Retrying...
 39 I0207 13:31:16.533837   12575 daemon.go:894] While getting MachineConfig master-cceb2fa93fab8800cb9e83839a1d0142, got: machineconfigs.machineconfiguration.openshift.io "master-cceb2fa93fab8800cb9e83839a1d0142" not found. Retrying...
 40 I0207 13:31:16.536906   12575 daemon.go:894] While getting MachineConfig master-cceb2fa93fab8800cb9e83839a1d0142, got: machineconfigs.machineconfiguration.openshift.io "master-cceb2fa93fab8800cb9e83839a1d0142" not found. Retrying...
 41 E0207 13:31:16.536948   12575 daemon.go:373] Fatal error checking initial state of node: Checking initial state: timed out waiting for the condition
 42 E0207 13:31:16.536983   12575 writer.go:85] Marking degraded due to: Checking initial state: timed out waiting for the condition
 43 I0207 13:31:16.552087   12575 daemon.go:375] Entering degraded state; going to sleep
```

**Comment by jlebon** [comment]
First two commits look good to me!

**Comment by runcom** [comment]
so, my very latest cluster is not exposing that master degradation anymore (using the very latest installer from master), let's see the CI here

File: test/e2e/main_test.go [hunk 1]
```diff
@@ -1,10 +1,60 @@
...
+	for _, cond := range clusterVersion.Status.Conditions {
+		if cond.Type == configv1.OperatorFailing && cond.Status == configv1.ConditionTrue {
+			cvoFailingErr = fmt.Errorf("CVO is failing with %q", cond.Message)
+			fmt.Println(cvoFailingErr)
+			break
+		}
+	}
+
+	if mcoFailingErr != nil && cvoFailingErr == nil {
+		panic("CVO is not failing but the MCO is failing, something went wrong")
```
Comment (cgwalters) [inline test/e2e/main_test.go]: In theory this is racy since the MCO could start failing right when we check, but...eh.  Not a real world concern I think.

**Comment by cgwalters** [comment]
Hm, looks like that last run failed with:

```
                  {
                        "lastTransitionTime": "2019-02-07T14:42:30Z",
                        "message": "Failed when progressing towards 3.11.0-584-g5cef4210-dirty because: error syncing: [timed out waiting for the condition during waitForControllerConfigToBeCompleted: controllerconfig is not completed: status for ControllerConfig machine-config-controller is being reported for 0, expecting it for 1, timed out waiting for the condition during syncRequiredMachineConfigPools: pool master has not progressed to latest configuration: configuration for pool master is empty]",
                        "reason": "error syncing: [timed out waiting for the condition during waitForControllerConfigToBeCompleted: controllerconfig is not completed: status for ControllerConfig machine-config-controller is being reported for 0, expecting it for 1, timed out waiting for the condition during syncRequiredMachineConfigPools: pool master has not progressed to latest configuration: configuration for pool master is empty]",
                        "status": "True",
                        "type": "Failing"
                    }
```

**Comment by runcom** [comment]
@cgwalters yeah, the CI is still failing but clusters spun up on the command line, from latest installer, doesn't expose that issue anymore :crossed_fingers:

File: test/e2e/main_test.go (hunk 1 above)
Comment (runcom) [reply test/e2e/main_test.go]: yeah, I'm adding this gate here just to be sure we are cool with 1) CVO being good 2) MCO being good just _before_ kicking e2e, but yeah, we could theoretically start failing afterwards

**Comment by runcom** [comment]
installer failing for some unknown reason:
```
time="2019-02-07T17:42:13Z" level=error msg="Error: Error applying plan:"
time="2019-02-07T17:42:13Z" level=error
time="2019-02-07T17:42:13Z" level=error msg="2 errors occurred:"
time="2019-02-07T17:42:13Z" level=error msg="\t* module.vpc.aws_route.to_nat_gw[4]: 1 error occurred:"
time="2019-02-07T17:42:13Z" level=error msg="\t* aws_route.to_nat_gw.4: Error creating route: timeout while waiting for state to become 'success' (timeout: 2m0s)"
time="2019-02-07T17:42:13Z" level=error
time="2019-02-07T17:42:13Z" level=error
time="2019-02-07T17:42:13Z" level=error msg="\t* module.vpc.aws_route.to_nat_gw[3]: 1 error occurred:"
time="2019-02-07T17:42:13Z" level=error msg="\t* aws_route.to_nat_gw.3: Error creating route: timeout while waiting for state to become 'success' (timeout: 2m0s)"
```
probably reached some limit on aws (???)

**Comment by runcom** [comment]
CI and AWS are already busted for today :sob: 
```
level=warning msg="Found override for ReleaseImage. Please be warned, this is not advised"
level=info msg="Consuming \"Install Config\" from target directory"
level=info msg="Creating cluster..."
level=error
level=error msg="Error: Error applying plan:"
level=error
level=error msg="1 error occurred:"
level=error msg="\t* module.vpc.aws_internet_gateway.igw: 1 error occurred:"
level=error msg="\t* aws_internet_gateway.igw: error attaching EC2 Internet Gateway (igw-05ed0db46fe3a3428): timeout while waiting for state to become 'success' (timeout: 2m0s)"
level=error
level=error
level=error
level=error
level=error
level=error msg="Terraform does not automatically rollback in the face of errors."
level=error msg="Instead, your Terraform state file has been partially updated with"
level=error msg="any resources that successfully completed. Please address the error"
level=error msg="above and apply again to incrementally change your infrastructure."
level=error
level=error
level=fatal msg="failed to fetch Cluster: failed to generate asset \"Cluster\": failed to create cluster: failed to apply using Terraform"
```

**Comment by runcom** [comment]
alright, this has started to pass now, latest commit named `x` bumps the timeout on the resources we're waiting on. @cgwalters @abhinavdahiya would you take another pass at this?

**Comment by abhinavdahiya** [comment]
[this](https://github.com/openshift/machine-config-operator/pull/386/commits/4108e90dfbb3528d7936ecb163e98e6fd08e5f11) is missing why we are checking for those conditions...
installer currently already waits for available CVO, that means available MCO...

[this](https://github.com/openshift/machine-config-operator/pull/386/commits/230aa5c5a84fbf741659a3d5becd5e534bf021a3) needs to have better title and message

otherwise looks good.

**Comment by runcom** [comment]
dropped https://github.com/openshift/machine-config-operator/pull/386/commits/4108e90dfbb3528d7936ecb163e98e6fd08e5f11 and reworded the last commit as well, let's see what the CI says now

**Comment by cgwalters** [comment]
> installer currently already waits for available CVO, that means available MCO...

But I keep circling back to the fact that clearly installs and e2e-aws in general were passing even if the MCO was in failing state.  Fixing *that* should be the primary goal - which *hopefully* this is doing?

**Comment by runcom** [comment]
> But I keep circling back to the fact that clearly installs and e2e-aws in general were passing even if the MCO was in failing state. Fixing _that_ should be the primary goal - which _hopefully_ this is doing?

yes, I had previously added a test commit which failed to converge to Available for the MCO and the CVO was indeed failing _correctly_

**Comment by cgwalters** [comment]
> installer currently already waits for available CVO, that means available MCO...

I'm going to cite this example again:
https://github.com/openshift/installer/pull/1189#issuecomment-460439928

That PR went through just fine, but the MCO had `Available: False` and `Failing: True`.  Now...it does seem likely that we only flipped to unavailable+failing after some time in available+progressing.  

So I think the important bit is really this part of the [CVO docs](https://github.com/openshift/cluster-version-operator/blob/master/docs/dev/clusteroperator.md)

> It then waits for the instance in the cluster until ... .status.conditions report available, not progressing and not failed

The "not progressing" part here being key.

**Comment by abhinavdahiya** [comment]
The previous error for [e2e-aws-op](https://openshift-gce-devel.appspot.com/build/origin-ci-test/pr-logs/pull/openshift_machine-config-operator/386/pull-ci-openshift-machine-config-operator-master-e2e-aws-op/360)
`Validate Response ec2/CreateVpc failed, not retrying, error RequestLimitExceeded: Request limit exceeded`
[openshift_install.log](https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_machine-config-operator/386/pull-ci-openshift-machine-config-operator-master-e2e-aws-op/360/artifacts/e2e-aws-op/installer/.openshift_install.log)
```console
time="2019-02-08T22:09:38Z" level=debug msg="2019-02-08T22:09:38.354Z [DEBUG] plugin.terraform-provider-aws: -----------------------------------------------------"
time="2019-02-08T22:09:38Z" level=debug msg="2019-02-08T22:09:38.497Z [DEBUG] plugin.terraform-provider-aws: 2019/02/08 22:09:38 [DEBUG] [aws-sdk-go] DEBUG: Response ec2/CreateVpc Details:"
time="2019-02-08T22:09:38Z" level=debug msg="2019-02-08T22:09:38.497Z [DEBUG] plugin.terraform-provider-aws: ---[ RESPONSE ]--------------------------------------"
time="2019-02-08T22:09:38Z" level=debug msg="2019-02-08T22:09:38.497Z [DEBUG] plugin.terraform-provider-aws: HTTP/1.1 503 Service Unavailable"
time="2019-02-08T22:09:38Z" level=debug msg="2019-02-08T22:09:38.497Z [DEBUG] plugin.terraform-provider-aws: Connection: close"
time="2019-02-08T22:09:38Z" level=debug msg="2019-02-08T22:09:38.497Z [DEBUG] plugin.terraform-provider-aws: Transfer-Encoding: chunked"
time="2019-02-08T22:09:38Z" level=debug msg="2019-02-08T22:09:38.497Z [DEBUG] plugin.terraform-provider-aws: Date: Fri, 08 Feb 2019 22:09:38 GMT"
time="2019-02-08T22:09:38Z" level=debug msg="2019-02-08T22:09:38.497Z [DEBUG] plugin.terraform-provider-aws: Server: AmazonEC2"
time="2019-02-08T22:09:38Z" level=debug msg="2019-02-08T22:09:38.497Z [DEBUG] plugin.terraform-provider-aws: "
time="2019-02-08T22:09:38Z" level=debug msg="2019-02-08T22:09:38.497Z [DEBUG] plugin.terraform-provider-aws: "
time="2019-02-08T22:09:38Z" level=debug msg="2019-02-08T22:09:38.497Z [DEBUG] plugin.terraform-provider-aws: -----------------------------------------------------"
time="2019-02-08T22:09:38Z" level=debug msg="2019-02-08T22:09:38.497Z [DEBUG] plugin.terraform-provider-aws: 2019/02/08 22:09:38 [DEBUG] [aws-sdk-go] <?xml version=\"1.0\" encoding=\"UTF-8\"?>"
time="2019-02-08T22:09:38Z" level=debug msg="2019-02-08T22:09:38.497Z [DEBUG] plugin.terraform-provider-aws: <Response><Errors><Error><Code>RequestLimitExceeded</Code><Message>Request limit exceeded.</Message></Error></Errors><RequestID>11dcca80-bd04-463e-9359-6dff5be0652e</RequestID></Response>"
time="2019-02-08T22:09:38Z" level=debug msg="2019-02-08T22:09:38.497Z [DEBUG] plugin.terraform-provider-aws: 2019/02/08 22:09:38 [DEBUG] [aws-sdk-go] DEBUG: Validate Response ec2/CreateVpc failed, not retrying, error RequestLimitExceeded: Request limit exceeded."
```

**Comment by runcom** [comment]
This is green now, I'm re-pushing to verify this again

**Comment by runcom** [comment]
Alrighty, all green

File: pkg/operator/sync.go [hunk 2]
```diff
@@ -329,7 +329,7 @@ func (optr *Operator) syncRequiredMachineConfigPools(config renderConfig) error
 		return err
 	}
 	var lastErr error
-	if err := wait.Poll(time.Second, 5*time.Minute, func() (bool, error) {
+	if err := wait.Poll(time.Second, 10*time.Minute, func() (bool, error) {
```
Comment (cgwalters) [inline pkg/operator/sync.go]: I feel like these timeouts may still be too short.  But we can revisit that later.

Comment (runcom) [reply pkg/operator/sync.go]: well, the installer waits up to 30 mins anyway, we could bump up to that if needed I guess

**Pull request merged**